linera-rpc/src/grpc/server.rs

```diff
@@ -399,75 +401,133 @@ where
                     }
 
-                    for i in 0..cross_chain_max_retries {
-                        // Delay increases linearly with the attempt number.
-                        linera_base::time::timer::sleep(
-                            cross_chain_sender_delay + cross_chain_retry_delay * i,
-                        )
-                        .await;
-
-                        let result = || async {
-                            let cross_chain_request = cross_chain_request.clone().try_into()?;
-                            let request = Request::new(cross_chain_request);
-                            let mut client =
-                                ValidatorWorkerClient::new(pool.channel(remote_address.clone())?)
-                                    .max_encoding_message_size(GRPC_MAX_MESSAGE_SIZE)
-                                    .max_decoding_message_size(GRPC_MAX_MESSAGE_SIZE);
-                            let response = client.handle_cross_chain_request(request).await?;
-                            Ok::<_, anyhow::Error>(response)
-                        };
-                        match result().await {
-                            Err(error) => {
-                                warn!(
-                                    nickname,
-                                    %error,
-                                    i,
-                                    from_shard = this_shard,
-                                    to_shard = shard_id,
-                                    "Failed to send cross-chain query",
-                                );
+                    let result = || async {
+                        let cross_chain_request = cross_chain_request.clone().try_into()?;
+                        let request = Request::new(cross_chain_request);
+                        let mut client =
+                            ValidatorWorkerClient::new(pool.channel(remote_address.clone())?)
+                                .max_encoding_message_size(GRPC_MAX_MESSAGE_SIZE)
+                                .max_decoding_message_size(GRPC_MAX_MESSAGE_SIZE);
+                        let response = client.handle_cross_chain_request(request).await?;
+                        Ok::<_, anyhow::Error>(response)
+                    };
+                    match result().await {
+                        Err(error) => {
+                            warn!(
+                                nickname,
+                                %error,
+                                i,
+                                from_shard = this_shard,
+                                to_shard = shard_id,
+                                "Failed to send cross-chain query",
+                            );
+                            i += 1;
+
+                            let mut guard = cross_chain_tasks.lock().unwrap();
+                            match guard.entry(cross_chain_request.sender_recipient_update()) {
```

Comment:
Nit: since you don't bind the vacant case you can just use `if let Some(entry) = guard.get_mut(…)`.